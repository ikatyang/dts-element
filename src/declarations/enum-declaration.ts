import * as ts from 'typescript';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { add_declare_modifier_if_need } from '../utils';
import { IVariableDeclaration } from './variable-declaration';

export interface IEnumDeclarationOptions extends IElementOptions {
  name: string;
  const?: boolean;
  export?: boolean;
  members?: IVariableDeclaration[];
}

export interface IEnumDeclaration
  extends IElement<ElementKind.EnumDeclaration>,
    IEnumDeclarationOptions {}

export const create_enum_declaration = (
  options: IEnumDeclarationOptions,
): IEnumDeclaration => ({
  ...create_element(ElementKind.EnumDeclaration),
  ...options,
});

export const is_enum_declaration = (value: any): value is IEnumDeclaration =>
  is_element(value) && value.kind === ElementKind.EnumDeclaration;

// tslint:disable:ter-indent

/**
 * @hidden
 */
export const transform_enum_declaration = (
  element: IEnumDeclaration,
  path: IElement<any>[],
) => {
  let counter = 0;

  const modifiers: ts.Modifier[] = [];
  if (element.export === true) {
    modifiers.push(ts.createToken(ts.SyntaxKind.ExportKeyword));
  }
  if (element.const === true) {
    modifiers.push(ts.createToken(ts.SyntaxKind.ConstKeyword));
  }

  return ts.createEnumDeclaration(
    /* decorators  */ undefined,
    /* modifiers   */ add_declare_modifier_if_need(modifiers, path),
    /* name        */ element.name,
    /* members     */ (element.members || []).map(variable_declaration => {
      if (
        variable_declaration.type &&
        variable_declaration.type.kind !== ElementKind.LiteralType
      ) {
        throw new Error(
          `enum_declaration.members[i].type should be a LiteralType`,
        );
      }
      let value = variable_declaration.type && variable_declaration.type.value;
      if (value === undefined) {
        value = counter++;
      } else if (typeof value === 'number') {
        counter = value + 1;
      }
      return ts.createEnumMember(
        /* name        */ variable_declaration.name,
        /* initializer */ ts.createLiteral(value),
      );
    }),
  );
};
