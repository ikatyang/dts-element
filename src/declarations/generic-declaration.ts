import * as ts from 'typescript';
import { IType } from '../collections';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { transform } from '../transform';

export interface IGenericDeclarationOptions extends IElementOptions {
  name: string;
  extends?: IType;
  defalut?: IType;
}

export interface IGenericDeclaration
  extends IElement<ElementKind.GenericDeclaration>,
    IGenericDeclarationOptions {}

export const create_generic_declaration = (
  options: IGenericDeclarationOptions,
): IGenericDeclaration => ({
  ...create_element(ElementKind.GenericDeclaration),
  ...options,
});

export const is_generic_declaration = (
  value: any,
): value is IGenericDeclaration =>
  is_element(value) && value.kind === ElementKind.GenericDeclaration;

/**
 * @hidden
 */
export const transform_generic_declaration = (
  element: IGenericDeclaration,
  path: IElement<any>[],
) =>
  ts.createTypeParameterDeclaration(
    /* name        */ element.name,
    /* constraint  */ element.extends &&
      (transform(element.extends, path) as ts.TypeNode),
    /* defaultType */ element.defalut &&
      (transform(element.defalut, path) as ts.TypeNode),
  );
