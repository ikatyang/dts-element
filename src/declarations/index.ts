// tslint:disable:max-line-length

export * from './class-declaration';
export * from './enum-declaration';
export * from './function-declaration';
export * from './generic-declaration';
export * from './global-declaration';
export * from './interface-declaration';
export * from './module-declaration';
export * from './namespace-declaration';
export * from './parameter-declaration';
export * from './type-declaration';
export * from './variable-declaration';
