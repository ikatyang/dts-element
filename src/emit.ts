import * as ts from 'typescript';
import { IElement } from './element';
import { transform } from './transform';

export const emit_native = (node: ts.Node, options?: ts.PrinterOptions) => {
  const source_file = ts.createSourceFile(
    '',
    '',
    ts.ScriptTarget.Latest,
    false,
    ts.ScriptKind.TS,
  );
  const printer = ts.createPrinter(options);
  return printer.printNode(ts.EmitHint.Unspecified, node, source_file);
};

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) =>
  emit_native(transform(element), options);
