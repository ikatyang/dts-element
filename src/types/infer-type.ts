import * as ts from 'typescript';
import { ElementKind } from '../constants';
import {
  transform_generic_declaration,
  IGenericDeclaration,
} from '../declarations/generic-declaration';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';

export interface IInferTypeOptions extends IElementOptions {
  generic: IGenericDeclaration;
}

export interface IInferType
  extends IElement<ElementKind.InferType>,
    IInferTypeOptions {}

export const create_infer_type = (options: IInferTypeOptions): IInferType => ({
  ...create_element(ElementKind.InferType),
  ...options,
});

export const is_infer_type = (value: any): value is IInferType =>
  is_element(value) && value.kind === ElementKind.InferType;

/**
 * @hidden
 */
export const transform_infer_type = (
  element: IInferType,
  path: IElement<any>[],
) =>
  ts.createInferTypeNode(
    /* typeParameter */ transform_generic_declaration(element.generic, path),
  );
