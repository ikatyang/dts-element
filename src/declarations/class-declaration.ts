import * as ts from 'typescript';
import { ElementKind } from '../constants';
import {
  create_element,
  is_element,
  IElement,
  IElementOptions,
} from '../element';
import { IClassMember } from '../members';
import { IIndexSignature } from '../others/index-signature';
import { transform } from '../transform';
import { IGeneralType } from '../types/general-type';
import {
  add_declare_modifier_if_need,
  create_expression_for_general_type,
  create_type_parameters,
} from '../utils';
import { IGenericDeclaration } from './generic-declaration';

export interface IClassDeclarationOptions extends IElementOptions {
  name: string;
  export?: boolean;
  generics?: IGenericDeclaration[];
  members?: (IClassMember | IIndexSignature)[];
  abstract?: boolean;
  extends?: IGeneralType;
}

export interface IClassDeclaration
  extends IElement<ElementKind.ClassDeclaration>,
    IClassDeclarationOptions {}

export const create_class_declaration = (
  options: IClassDeclarationOptions,
): IClassDeclaration => ({
  ...create_element(ElementKind.ClassDeclaration),
  ...options,
});

export const is_class_declaration = (value: any): value is IClassDeclaration =>
  is_element(value) && value.kind === ElementKind.ClassDeclaration;

// tslint:disable:ter-indent max-line-length

/**
 * @hidden
 */
export const transform_class_declaration = (
  element: IClassDeclaration,
  path: IElement<any>[],
) =>
  ts.createClassDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
      [
        ...(element.export === true
          ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
          : []),
        ...(element.abstract === true
          ? [ts.createToken(ts.SyntaxKind.AbstractKeyword)]
          : []),
      ],
      path,
    ),
    /* name            */ element.name,
    /* typeParameters  */ create_type_parameters(element.generics, path),
    /* heritageClauses */ element.extends === undefined
      ? []
      : [
          ts.createHeritageClause(
            /* token */ ts.SyntaxKind.ExtendsKeyword,
            /* types */ [
              create_expression_for_general_type(element.extends, path),
            ],
          ),
        ],
    /* members         */ (element.members || [])
      .map(member => transform(member, path) as ts.ClassElement),
  );
