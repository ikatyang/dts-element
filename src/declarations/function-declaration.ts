import * as ts from 'typescript';
import { ElementKind } from '../constants';
import { create_element, IElement, IElementOptions } from '../element';
import { transform } from '../transform';
import { create_function_type, IFunctionType } from '../types/function-type';
import { add_declare_modifier_if_need } from '../utils';

export interface IFunctionDeclarationOptions extends IElementOptions {
  name: string | undefined;
  type?: IFunctionType;
  export?: boolean;
}

export interface IFunctionDeclaration
  extends IElement<ElementKind.FunctionDeclaration>,
    IFunctionDeclarationOptions {}

export const create_function_declaration = (
  options: IFunctionDeclarationOptions,
): IFunctionDeclaration => ({
  ...create_element(ElementKind.FunctionDeclaration),
  ...options,
});

// tslint:disable:ter-indent

/**
 * @hidden
 */
export const transform_function_declaration = (
  element: IFunctionDeclaration,
  path: IElement<any>[],
) => {
  const function_type = transform(
    element.type || create_function_type(),
    path,
  ) as ts.FunctionDeclaration;
  return ts.createFunctionDeclaration(
    /* decorators      */ undefined,
    /* modifiers       */ add_declare_modifier_if_need(
      element.export === true
        ? [ts.createToken(ts.SyntaxKind.ExportKeyword)]
        : undefined,
      path,
    ),
    /* asteriskToken   */ undefined,
    /* name            */ element.name,
    /* typeParameters  */ function_type.typeParameters,
    /* parameters      */ function_type.parameters,
    /* type            */ function_type.type,
    /* body            */ undefined,
  );
};
