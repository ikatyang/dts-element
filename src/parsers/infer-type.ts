import * as ts from 'typescript';
import { IGenericDeclaration } from '../declarations/generic-declaration';
import { parse_native } from '../parse';
import { create_infer_type, IInferType } from '../types/infer-type';

export const parse_infer_type = (node: ts.InferTypeNode): IInferType =>
  create_infer_type({
    generic: parse_native(node.typeParameter) as IGenericDeclaration,
  });
