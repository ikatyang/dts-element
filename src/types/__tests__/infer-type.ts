import { create_generic_declaration } from '../../declarations/generic-declaration';
import { emit } from '../../emit';
import { create_infer_type, is_infer_type } from '../infer-type';

it('should return correctly', () => {
  expect(
    emit(
      create_infer_type({
        generic: create_generic_declaration({ name: 'T' }),
      }),
    ),
  ).toMatchSnapshot();
});

describe('is_infer_type', () => {
  it('should return correctly', () => {
    const element = create_infer_type({
      generic: create_generic_declaration({ name: 'T' }),
    });
    expect(is_infer_type(element)).toBe(true);
  });
});
