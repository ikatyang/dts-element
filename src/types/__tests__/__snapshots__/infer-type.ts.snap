// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `"infer T"`;
